availability/etcdfailover/electionmanager.go

```diff
@@ -0,0 +1,234 @@
+		em.evtChan = NewSafeSendChan()
+		var err error
+		//ensure consecutive start-ups don't happen
+		defer func() {
+			em.setState(Participant)
+		}()
+
+		if em.session == nil {
+			if em.cli == nil {
+				em.cliCreated = true
+				em.cli, err = clientv3.New(em.cfg.Etcd)
+				if err != nil {
+					return nil, fmt.Errorf("failed to setup etcd client: %v", err)
+				}
+				em.cfg.Logger.Debug("etcd client created")
+			}
+			em.sessionCreated = true
+			em.session, err = concurrency.NewSession(em.cli, concurrency.WithTTL(em.cfg.SessionTimeoutSecs))
+			if err != nil {
+				return nil, fmt.Errorf("failed to create session: %v", err)
+			}
+			em.cfg.Logger.Debug("etcd session created")
+		}
+
+		em.election = concurrency.NewElection(em.session, em.cfg.ElectionPrefix)
+		em.cfg.Logger.Debug("etcd election created")
+
+		var cctx context.Context
+		cctx, em.cancel = context.WithCancel(context.Background())
+
+		go func() {
+			em.cfg.Logger.Debug("observing etcd session expiry")
+			<-em.session.Done()
+			em.cfg.Logger.Error("etcd session expired")
+			em.StopAndLog()
+		}()
+
+		go func() {
+			em.cfg.Logger.Debug("starting election campaign")
+			if err := em.election.Campaign(cctx, ""); err != nil {
+				em.cfg.Logger.Errorf("failed during election campaign: %v", err)
+				em.StopAndLog()
+			} else {
+				em.setState(Leader)
+			}
+			em.cfg.Logger.Debug("stopped election campaign")
+		}()
+
+		return em.evtChan.RecvChan(), nil
+	} else {
+		return nil, fmt.Errorf("cannot setup election manager. already started, state = %s", em.state)
+	}
+}
+
+func (em *ElectionManager) Stop() <-chan error {
+	c := make(chan error)
+
+	go func() {
+		em.mu.Lock()
+		defer em.mu.Unlock()
+		if err := em.stop(); err != nil {
+			c <- err
+			return
+		}
+		close(c)
+	}()
+
+	return c
+}
+
+func (em *ElectionManager) StopAndLog() {
+	serr := <-em.Stop()
+	if serr != nil {
+		em.cfg.Logger.Warnf("error on stopping election manager: %v", serr)
+	}
+}
+
+func (em *ElectionManager) Run(runnable func() error) (<-chan error, func() <-chan error, error) {
+	em.cfg.Logger.Info("starting election participation")
+	var (
+		electionChan <-chan ElectionState
+		err	error
+	)
+	if electionChan, err = em.Start(); err != nil {
+		em.StopAndLog()
+		return nil, nil, err
+	}
+
+	errChan := make(chan error)
+	go func(runnable func() error) {
+		em.cfg.Logger.Debug("waiting for election event or quit signal")
+		for evt := range electionChan {
+			switch evt {
+			case Participant:
+				em.cfg.Logger.Info("tracker is eligible to become leader")
+			case Leader:
+				em.cfg.Logger.Info("leader among trackers")
+				if err = runnable(); err != nil {
+					em.cfg.Logger.Errorf("encountered error in runnable: %v", err)
+					errChan <- err
+				}
+			}
+		}
+		errChan <- fmt.Errorf("failed in election")
+	}(runnable)
+
+	return errChan, em.Stop, nil
+}
+
+//method not thread-safe, ensure it is called after acquiring mutex of election manager
+func (em *ElectionManager) stop() error {
```

Comment:
Might be wrong here, but I guess you can also get rid of `SafeSendChan` if you wait for 2 goutines(spawned in `Start`), and close the chan after that.

Reply:
nah, stop waiting on them is not necessary if they are guaranteed to exit.
the goroutine monitoring election campaign is guaranteed to exit since we do resign/lease expiry in stop. although it will unnecessarily call stopandlog again (no side-effects). I have cleaned that up regardless so now if context is cancelled that goroutine will not call stop again.
the goroutine monitoring session expiry is also guaranteed to exit if this was a session created by election manager itself since it is closed in Stop method. but yes, this goroutine would be a leak and cause correctness issues if session is provided by user. fixed that.

Reply:
I agree. corner scenario definitely exists. made changes to address it.